pkg/controller/admissionchecks/multikueue/fswatch.go

```diff
@@ -0,0 +1,226 @@
+	reconcile        chan event.GenericEvent
+}
+
+var _ manager.Runnable = (*KubeConfigFSWatcher)(nil)
+
+func newKubeConfigFSWatcher() *KubeConfigFSWatcher {
+	return &KubeConfigFSWatcher{
+		clusterToFile:    map[string]string{},
+		fileToClusters:   map[string]set.Set[string]{},
+		parentDirToFiles: map[string]set.Set[string]{},
+		reconcile:        make(chan event.GenericEvent),
+	}
+}
+
+func (w *KubeConfigFSWatcher) Start(ctx context.Context) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	var err error
+	w.watcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		log := ctrl.LoggerFrom(ctx).WithName("kubeconfig-fs-watcher")
+		defer func() {
+			err := w.watcher.Close()
+			if err != nil {
+				log.V(2).Error(err, "Closing kubeconfigs FS Watcher")
+			}
+		}()
+		watchedEvents := fsnotify.Write | fsnotify.Create | fsnotify.Remove | fsnotify.Rename
+		for {
+			select {
+			case ev := <-w.watcher.Events:
+				log.V(5).Info("Got event", "name", ev.Name, "op", ev.Op)
+				switch {
+				case (ev.Op & watchedEvents) != 0:
+					w.notifyPathWrite(ev.Name)
+				default:
+					log.V(2).Info("Ignore event", "name", ev.Name, "op", ev.Op)
+
+				}
+			case err := <-w.watcher.Errors:
+				log.V(2).Error(err, "Kubeconfigs FS Watch")
+			case <-ctx.Done():
+				log.V(2).Info("End kubeconfigs FS Watch")
+				return
+			}
+		}
+
+	}()
+	return nil
+}
+
+func (w *KubeConfigFSWatcher) notifyPathWrite(path string) {
+	w.lock.RLock()
+	clusters := w.fileToClusters[path].UnsortedList()
+	w.lock.RUnlock()
+	//TODO: maybe manage "..data" pattern
+	for _, c := range clusters {
+		w.reconcile <- event.GenericEvent{Object: &kueuealpha.MultiKueueCluster{ObjectMeta: metav1.ObjectMeta{Name: c}}}
+	}
+}
+
+func (w *KubeConfigFSWatcher) getClusterPath(cluster string) (string, bool) {
+	w.lock.RLock()
+	defer w.lock.RUnlock()
+	path, found := w.clusterToFile[cluster]
+	return path, found
+}
+
+func (w *KubeConfigFSWatcher) set(cluster, kcPath string) error {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	dir := path.Dir(kcPath)
+	if _, found := w.parentDirToFiles[dir]; !found {
+		err := w.watcher.Add(dir)
+		if err != nil {
+			return err
+		}
+	}
+
+	w.clusterToFile[cluster] = kcPath
+	if _, found := w.fileToClusters[kcPath]; found {
+		w.fileToClusters[kcPath].Insert(cluster)
+	} else {
+		w.fileToClusters[kcPath] = set.New(cluster)
+	}
+	if _, found := w.parentDirToFiles[dir]; found {
+		w.parentDirToFiles[dir].Insert(kcPath)
+	} else {
+		w.parentDirToFiles[dir] = set.New(kcPath)
+		return w.watcher.Add(dir)
+	}
+	return nil
+}
+
+func (w *KubeConfigFSWatcher) remove(cluster string) {
+	w.lock.Lock()
+	defer w.lock.Unlock()
+
+	kcPath := w.clusterToFile[cluster]
+	dir := path.Dir(kcPath)
+
+	file_removed := false
+	if s, found := w.fileToClusters[kcPath]; found {
+		s.Delete(cluster)
+		if s.Len() == 0 {
+			delete(w.fileToClusters, kcPath)
+			file_removed = true
+		}
+	}
+
+	if file_removed {
```

Comment:
If the file is created manually by an admin, I don't think we need to delete it in Kueue. I would suggest leave it. Makes less code, and gives full responsibility to the admin.

Reply:
I'm also concerned that there might be a risk that a user could by mistake, or maliciously, misconfigure the file, and then Kueue removes a potentially important file (pointed by the path).

Reply:
Gotcha, thanks for explaining.